{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2ee4d388",
   "metadata": {},
   "source": [
    "# HEADER HERE\n",
    "Text\n",
    "- bullet points"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "928357af",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Script to get folding energies and transcript sequences using genes from DESeq2 output\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import RNA\n",
    "\n",
    "#Use tx2gene file used for DESeq2, since DESeq2 output is in common name rather than SGD IDs\n",
    "TxToGeneFile = \"tx2gene.txt\"\n",
    "\n",
    "#Use fasta file generated using SGD R64-1-1 CDSes and Pelechano UTR annotation\n",
    "#Uses longest UTR if available or default 5' and 3'UTRs of 120 nt and 200 nt, respectively, if none found\n",
    "TranscriptomeFile = \"R64-1-1_transcriptome_with_utrs.fasta\"\n",
    "\n",
    "#return dictionary of transcript sequences mapped to common names\n",
    "transcripts = {}\n",
    "\n",
    "#Build dictionary of common names mapped to SGD IDs\n",
    "txToGene = {}\n",
    "with open(txToGeneFile, \"r\") as inFile:\n",
    "        for line in conversionFile:\n",
    "            #Skip the first line\n",
    "            if (\"GENEID\" not in line):\n",
    "                data = line.strip().split(\"\\t\")\n",
    "                transcriptId = data[0]\n",
    "                geneId = data[1]\n",
    "                txToGene[geneId] = transcriptId\n",
    "            \n",
    "    #Build dictionary of transcript sequences mapped to SGD ID\n",
    "    transcripts = {}\n",
    "    transcriptFasta = SeqIO.parse(open(transcriptFile), 'fasta')\n",
    "    for fasta in transcriptFasta:\n",
    "        name = geneToTx()\n",
    "        sequence = str(fasta.seq)\n",
    "        if (asRNA):\n",
    "            sequence = sequence.replace(\"T\", \"U\")\n",
    "        transcripts[fasta.name] = sequence\n",
    "        \n",
    "    return transcripts\n",
    "\n",
    "def getStremeFasta(geneList, outputFile, asRNA = True, txToGeneFile = defaultTxToGene, transcriptFile = defaultTranscriptome):\n",
    "    transcriptList = genesToTranscripts(asRNA, txToGeneFile, transcriptFile)\n",
    "    \n",
    "    #Parse gene list to write out transcript sequences\n",
    "    with open(geneList, \"r\") as inFile, open(outputFile, \"w\") as outFile:\n",
    "        for line in inFile:\n",
    "            gene = line.strip()\n",
    "            #convert common name to SGD ID\n",
    "            sgdId = transcriptList[gene]\n",
    "            #get transcript sequence\n",
    "            transcript = transcripts[sgdId]\n",
    "            #write to fasta\n",
    "            outFile.write(\">\" + sgdId + \"\\n\")\n",
    "            outFile.write(transcript + \"\\n\")\n",
    "    \n",
    "        #Get gene list to write out transcript sequences\n",
    "    with open(geneList, \"r\") as inFile, open(outputFile, \"w\") as outFile:\n",
    "        for line in inFile:\n",
    "            gene = line.strip()\n",
    "            #convert common name to SGD ID\n",
    "            sgdId = geneToTx[gene]\n",
    "            #get transcript sequence\n",
    "            transcript = transcripts[sgdId]\n",
    "            #write to fasta\n",
    "            outFile.write(\">\" + sgdId + \"\\n\")\n",
    "            outFile.write(transcript + \"\\n\")\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    #Build dictionary of transcript sequences for the given gene list\n",
    "    outputTranscripts = {}\n",
    "    with open(geneList, \"r\") as inFile:\n",
    "        for line in inFile:\n",
    "            gene = line.strip()\n",
    "            #convert common name to SGD ID\n",
    "            sgdId = geneToTx[gene]\n",
    "            transcriptSeq = transcripts[sgdId]\n",
    "            if (asRNA):\n",
    "                transcriptSeq = transcriptSeq.replace(\"T\", \"U\")\n",
    "            outputTranscripts[sgdId] = transcriptSeq\n",
    "            \n",
    "    return transcripts\n",
    "        \n",
    "def getStremeFasta(geneList, asRNA = True, txToGeneFile = defaultTxToGene, transcriptFile = defaultTranscriptome):\n",
    "        #Get gene list to write out transcript sequences\n",
    "    with open(geneList, \"r\") as inFile, open(outputFile, \"w\") as outFile:\n",
    "        for line in inFile:\n",
    "            gene = line.strip()\n",
    "            #convert common name to SGD ID\n",
    "            sgdId = geneToTx[gene]\n",
    "            #get transcript sequence\n",
    "            transcript = transcripts[sgdId]\n",
    "            #write to fasta\n",
    "            outFile.write(\">\" + sgdId + \"\\n\")\n",
    "            outFile.write(transcript + \"\\n\")\n",
    "            \n",
    "    \n",
    "def getFoldingEnergies(geneList, asRNA = True, txToGeneFile = defaultTxToGene, transcriptFile = defaultTranscriptome):\n",
    "        #Get gene list to write out transcript sequences\n",
    "    with open(geneList, \"r\") as inFile, open(outputFile, \"w\") as outFile:\n",
    "        for line in inFile:\n",
    "            gene = line.strip()\n",
    "            #convert common name to SGD ID\n",
    "            sgdId = geneToTx[gene]\n",
    "            #get transcript sequence\n",
    "            transcript = transcripts[sgdId]\n",
    "            #write to fasta\n",
    "            outFile.write(\">\" + sgdId + \"\\n\")\n",
    "            outFile.write(transcript + \"\\n\")\n",
    "            \n",
    "    \n",
    "    with open(\"tx2gene.txt\", \"r\") as tx2gene_file, open(\"TE Multivariate List.txt\", \"r\") as in_file:\n",
    "    for line in tx2gene_file:\n",
    "        if (\"GENEID\" not in line):\n",
    "            data = line.strip().split(\"\\t\")\n",
    "            tx2gene[data[1]] = data[0]\n",
    "            \n",
    "    for line in in_file:\n",
    "        genes.append(tx2gene[line.strip()])\n",
    "\n",
    "transcriptome = {}\n",
    "with open(\"TE_dG.txt\", \"w\") as out_file:        \n",
    "    transcript_file = SeqIO.parse(open(\"R64-1-1_transcriptome_with_utrs.fasta\"), 'fasta')\n",
    "    for fasta in transcript_file:\n",
    "        sequence = str(fasta.seq)\n",
    "        sequence = sequence.replace(\"T\", \"U\")\n",
    "        transcriptome[fasta.name] = sequence\n",
    "        \n",
    "    for gene in genes:\n",
    "        (ss, mfe) = RNA.fold(transcriptome[gene])\n",
    "        out_file.write(gene + \"\\t\" + str(mfe) + \"\\n\")\n",
    "    \n",
    "\n",
    "\n",
    "def getStremeFasta(txToGeneFile, geneList, transcriptFile, outputFile):\n",
    "    \n",
    "    #Build dictionary of SGD IDs mapped to common names\n",
    "    geneToTx = {}\n",
    "    with open(txToGeneFile, \"r\") as conversionFile:\n",
    "        for line in conversionFile:\n",
    "            #Skip the first line\n",
    "            if (\"GENEID\" not in line):\n",
    "                data = line.strip().split(\"\\t\")\n",
    "                geneToTx[data[1]] = data[0]\n",
    "            \n",
    "    #Build dictionary of transcript sequences mapped to SGD ID\n",
    "    transcripts = {}\n",
    "    transcriptFasta = SeqIO.parse(open(transcriptFile), 'fasta')\n",
    "    for fasta in transcriptFasta:\n",
    "        sequence = str(fasta.seq)\n",
    "        #replace all T's with U's to convert DNA to RNA\n",
    "        sequence = sequence.replace(\"T\", \"U\")\n",
    "        transcripts[fasta.name] = sequence\n",
    "        \n",
    "    #Get gene list to write out transcript sequences\n",
    "    with open(geneList, \"r\") as inFile, open(outputFile, \"w\") as outFile:\n",
    "        for line in inFile:\n",
    "            gene = line.strip()\n",
    "            #convert common name to SGD ID\n",
    "            sgdId = geneToTx[gene]\n",
    "            #get transcript sequence\n",
    "            transcript = transcripts[sgdId]\n",
    "            #write to fasta\n",
    "            outFile.write(\">\" + sgdId + \"\\n\")\n",
    "            outFile.write(transcript + \"\\n\")\n",
    "            \n",
    "\n",
    "#Get fasta file for TEs\n",
    "getStremeFasta(defaultTxToGene, \"Up in TE.txt\", defaultTranscriptome, \"Up in TE.fasta\")\n",
    "#Get fasta file for ribosome profiling\n",
    "getStremeFasta(defaultTxToGene, \"Up in RPF.txt\", defaultTranscriptome, \"Up in RPF.fasta\")\n",
    "#Get fasta file for RNA-seq\n",
    "getStremeFasta(defaultTxToGene, \"Up in RNA.txt\", defaultTranscriptome, \"Up in RNA.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20fe7637",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Script to get folding energies for transcripts using ViennaRNA RNAFold\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import RNA\n",
    "\n",
    "def getFoldingEnergies(txToGeneFile, geneList, transcriptFile, outputFile):\n",
    "    \n",
    "\n",
    "\n",
    "tx2gene = {}\n",
    "genes = []\n",
    "\n",
    "with open(\"tx2gene.txt\", \"r\") as tx2gene_file, open(\"TE Multivariate List.txt\", \"r\") as in_file:\n",
    "    for line in tx2gene_file:\n",
    "        if (\"GENEID\" not in line):\n",
    "            data = line.strip().split(\"\\t\")\n",
    "            tx2gene[data[1]] = data[0]\n",
    "            \n",
    "    for line in in_file:\n",
    "        genes.append(tx2gene[line.strip()])\n",
    "\n",
    "transcriptome = {}\n",
    "with open(\"TE_dG.txt\", \"w\") as out_file:        \n",
    "    transcript_file = SeqIO.parse(open(\"R64-1-1_transcriptome_with_utrs.fasta\"), 'fasta')\n",
    "    for fasta in transcript_file:\n",
    "        sequence = str(fasta.seq)\n",
    "        sequence = sequence.replace(\"T\", \"U\")\n",
    "        transcriptome[fasta.name] = sequence\n",
    "        \n",
    "    for gene in genes:\n",
    "        (ss, mfe) = RNA.fold(transcriptome[gene])\n",
    "        out_file.write(gene + \"\\t\" + str(mfe) + \"\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
